kimchi/src/snarky/ec.rs

```diff
@@ -0,0 +1,304 @@
+//! Functions associated to the EC addition and scaling.
+
+use super::{
+    constraint_system::{EcAddCompleteInput, ScaleRound},
+    snarky_type::SnarkyType,
+};
+use crate::{
+    snarky::{
+        constraint_system::KimchiConstraint,
+        prelude::{FieldVar, RunState},
+        runner::Constraint,
+    },
+    SnarkyResult,
+};
+use ark_ec::{short_weierstrass_jacobian::GroupAffine, SWModelParameters};
+use ark_ff::{BigInteger, PrimeField};
+use std::{borrow::Cow, marker::PhantomData};
+
+pub fn ec_add<F: PrimeField, C: SWModelParameters<BaseField = F>>(
+    runner: &mut RunState<F>,
+    loc: Cow<'static, str>,
+    a: [FieldVar<F>; 2],
+    b: [FieldVar<F>; 2],
+) -> SnarkyResult<[FieldVar<F>; 2]> {
+    let [a0, a1] = a;
+    let [b0, b1] = b;
+    let [p3x, p3y, inf, same_x, slope, inf_z, x21_inv]: [FieldVar<F>; 7] =
+        runner.compute(loc.clone(), |w| {
+            let to_bool = |x| {
+                if x {
+                    F::from(1_u8)
+                } else {
+                    F::zero()
+                }
+            };
+            let x1 = w.read_var(&a0);
+            let x2 = w.read_var(&a1);
+            let y1 = w.read_var(&b0);
+            let y2 = w.read_var(&b1);
+
+            let same_x = to_bool(x1 == x2);
+
+            let x_diff = x2 - x1;
+            let x21_inv = x_diff.inverse().unwrap();
```

Comment:
So the actual constraint is something like `x21_inv * x_diff = 1`, right?

Reply:
slightly more complex, you can see in https://o1-labs.github.io/proof-systems/specs/kimchi.html#elliptic-curve-addition that is used to constrain `same_x`.